### ISSUE TITLE
Inconsistency in locking behavior between AOCSCompact() vs. AppendOnlyCompact()

### ISSUE THREAD

**Issue by fangzheng** [greenplum-db/gpdb#10529]
### Greenplum version or build
6X_STABLE

In 6X_STABLE branch, both `AOCSCompact()` and `AppendOnlyCompact()` functions call `LockRelationAppendOnlySegmentFile()`, but they set the `dontWait` parameter differently. `AOCSCompact()` sets `dontWait` to true and skips any seg file that has been locked (https://github.com/greenplum-db/gpdb/blob/6X_STABLE/src/backend/access/aocs/aocs_compaction.c#L558), whereas `AppendOnlyCompact()` sets `dontWait` to false (https://github.com/greenplum-db/gpdb/blob/6X_STABLE/src/backend/access/appendonly/appendonly_compaction.c#L715). I am wondering if there is any potential locking issue.

**Comment by asimrp** [comment]
Indeed, that's a good find @fangzheng.  Let me see if that could cause a problem, especially with an insert and vacuum running concurrently.

**Comment by asimrp** [comment]
@bbbearxyz @wraymo this is issue is not related directly but seems worth looking at after you have finished PR #11895.  FYI @kainwen

**Comment by huansong** [comment]
From consistence perspective it's OK:
* In AOCSCompact(), we don't wait on the lock but checks if the lock is acquired or not. If acquired successfully, go ahead and do the rest. If not, we skip that segfile and continue.
* In AppendOnlyCompact(), we block till we acquired the lock.

So in either case, vacuuming is guaranteed to be exclusive with any concurrent operations on the segfiles. The difference is just that vacuuming AO table will block wait but vacuuming AOCO may skip some segfile. 

However, since we are silently skipping some segfiles is that a correctness concern? It would appear to the user that we vacuum'ed the table w/o an issue even though we did not. We probably should either fix or document this behavior.

P.S. in 7X, it is changed to lock the corresponding pg_aoseg row instead of the relation: [commit](https://github.com/greenplum-db/gpdb/commit/5778f31124d5#diff-9fecdbbedbe0a8ed6898b84ef38c28e23ae4a8f4d6f09b38f57be565c4ee8504). When acquiring the lock, it will skip any tuple that it cannot acquire lock (`choose_segno_internal`), but ultimately it keeps looping through all segments necessary (`ao_vacuum_rel_compact`), so it's effectively same as block wait. And that's the same for **both** AO and AOCO tables.

**Comment by huansong** [comment]
From correctness perspective it still seems to be OK:

* First of all, as mentioned, the only case we care about is concurrent INSERT during the compact phase of vacuum.
  * For other queries that affects the content of the segfiles, they will need either AccessExclusiveLock or ShareUpdateExclusiveLock on the table, both of which would conflict with what the VACUUM holds on the table. So they cannot happen concurrently.
  * For the other vacuum phases, the problem in AOCSCompact() won't have an impact.
* Now let's consider the cases for concurrent INSERT during compact:
  * If the INSERT comes before we choose the next segfile to compact (`vacuum_assign_compaction_segno`), the segfile that the INSERT is using won't be selected. So no conflict would happen in AOCSCompact().
  * If the INSERT comes after we choosed any segfile to compact (and before we do the compact), the INSERT would select another segfile because we've marked the compact-able segfile in-use already. So no conflict there either.

Therefore, it seems that we don't need to do anything for this, besides an attempt to make AO and AOCO code exactly the same, which I feel doesn't worth the effort. Closing the issue as is for now.
